rust/src/http2/http2.rs

```diff
@@ -0,0 +1,1175 @@
+                    //we may have consumed data from previous records
+                    if input.len() < HTTP2_FRAME_HEADER_LEN {
+                        return AppLayerResult::incomplete(
+                            (il - input.len()) as u32,
+                            HTTP2_FRAME_HEADER_LEN as u32,
+                        );
+                    } else {
+                        panic!("HTTP2 invalid length frame header");
+                    }
+                }
+                Err(_) => {
+                    self.set_event(HTTP2Event::InvalidFrameHeader);
+                    return AppLayerResult::err();
+                }
+            }
+        }
+        return AppLayerResult::ok();
+    }
+
+    fn parse_tc(&mut self, mut input: &[u8]) -> AppLayerResult {
+        //first consume frame bytes
+        let il = input.len();
+        if self.response_frame_size > 0 {
+            let ilen = input.len() as u32;
+            if self.response_frame_size >= ilen {
+                self.response_frame_size -= ilen;
+                return AppLayerResult::ok();
+            } else {
+                let start = self.response_frame_size as usize;
+                input = &input[start..];
+                self.response_frame_size = 0;
+            }
+        }
+        //then parse all we can
+        while input.len() > 0 {
+            match parser::http2_parse_frame_header(input) {
+                Ok((rem, head)) => {
+                    let hl = head.length as usize;
+
+                    //we check for completeness first
+                    if rem.len() < hl {
+                        //but limit ourselves so as not to exhaust memory
+                        if hl < HTTP2_MAX_HANDLED_FRAME_SIZE {
+                            return AppLayerResult::incomplete(
+                                (il - input.len()) as u32,
+                                (HTTP2_FRAME_HEADER_LEN + hl) as u32,
+                            );
+                        } else if rem.len() < HTTP2_MIN_HANDLED_FRAME_SIZE {
+                            return AppLayerResult::incomplete(
+                                (il - input.len()) as u32,
+                                (HTTP2_FRAME_HEADER_LEN + HTTP2_MIN_HANDLED_FRAME_SIZE) as u32,
+                            );
+                        } else {
+                            self.set_event(HTTP2Event::LongFrameData);
+                            self.request_frame_size = head.length - (rem.len() as u32);
+                        }
+                    }
+
+                    //get a safe length for the buffer
+                    let (hlsafe, complete) = if rem.len() < hl {
+                        (rem.len(), false)
+                    } else {
+                        (hl, true)
+                    };
+
+                    if head.length == 0 && head.ftype == parser::HTTP2FrameType::SETTINGS as u8 {
+                        input = &rem[hlsafe..];
```

Comment:
I see no explicit check for hlsafe (typo? hlsave?) to be > 0. Concern is potential for endless loops

Reply:
Even if `hlsafe` = 0 (no typo, it is meant to be safe), there should be no buffer overflow as `rem` should start after `input` (the 9 bytes of the HTTP2 frame header)